Fix: #1675 - Console User Interface [WIP]

raygervais opened the pull request:
This work-in-progress PR implements the console interface enhancement originally discussed [here](https://github.com/mozilla/thimble.mozilla.org/issues/1675).  I've also included my previous observations in the description for context and conceptualization. 

@flukeout I was recommended to ask you how you'd like the console styled, including text colours, what arguments or data should be included (I'm thinking time stamps), font-faces, and layout as well. 

Dave was concerned that it spanning the entire bottom may take too much room on lower-resolution devices. Furthermore, I'm aware that I have to implement a `hide()` function for when the user doesn't wish to see the console, which would mean perhaps we had a button to call the `show()` function. Did you have any ideas where this button should be placed if we were to use it? 

I also wanted to ask you about the idea of keeping the console in the editor-frame alone, so that the live preview was unobstructed. Opinion? 

![screenshot from 2017-03-28 15-02-17](https://cloud.githubusercontent.com/assets/14265337/24422491/b50ec7a6-13c7-11e7-87c2-8c522c61b739.png)

Another idea that I presented Dave with, was reducing the console's frame to the very bottom, showing only the console tool bar and perhaps instead of an close button, an open button. This would eliminate the **live preview only** button, which clears and then freezes the console.

![image](https://cloud.githubusercontent.com/assets/14265337/24422652/3bf22678-13c8-11e7-8f0b-0fd17f2c03af.png)

There is also a debate on separating the various `console.*` functions as they did above, so that you can filter and view the ones you wish to, or if we should let all the functions display in a single pane without the filtering or buttons.

raygervais commented:
@humphd, if you'd like I can rebase to clean up the commits since many are from the back-end implementation, which I forked from.

humphd commented:
Yes, please rebase your work on `master`, since your other code has landed there.

flukeout commented:
Hey this is really awesome @raygervais! Got it up and running here's a few questions:

* What does "Live Preview Only" mean
  * The console only seems to work when this option is enabled
* There are lots of console entries that seem to related to the brackets environment, and not the users code, for example...
  * ``Possible memory leak: 19 'StylesheetRemoved' listeners attached to``
  * ``Deleted /2/projects/36/Test/index.html Document still has 1 referen... ``
  * Can we filter these out so that we only report on the user's code?

raygervais commented:
Thank you @flukeout! I noticed that the button's functionality for "Life Preview Only" is attempting to remove focus on the console, as if the implementation was never supposed to minimize. We could remove that once you address how you'd like for me to handle when the console should and shouldn't be visible. 

I've implemented the back end, which means I just need to hook it up to the `ConsoleInterfaceManager.js` file for only user-based code only to be reported. I will work on that after cleaning up my git history above. 

Sorry for the mess above @humphd, I may bug you if I cannot figure it out later this week for a quick git rebase refresher. I've managed to rebase my previous PR's branch into `master`, but for some reason I can't rebase this branch from `master` after. Ideas?

flukeout commented:
@raygervais For now, let's add a new element to the page like this...

![image](https://cloud.githubusercontent.com/assets/25212/24479535/a5905326-1494-11e7-9b27-410ce75db61e.png)

Let's give it fixed positioning and place it in the bottom right corner of the screen. When this is clicked, let's hide it and show the console.

Maybe down the road we can put it into the Preview bar, but I think for now this will be fine. It can just be a ``<div>Console</div>`` element for now. Ping me when you make another push and I can start working on styling.

flukeout commented:
Another question

If the console is closed, would it be possible to keep track of how many unseen console log items there are and display it somewhere?

![image](https://cloud.githubusercontent.com/assets/25212/24479692/4e2ac034-1495-11e7-89b7-9a89d21f7e69.png)

Not a requirement at all, but might be a nice feature.

humphd commented:
What do you think of only showing that button when there actually are console logs to see?  There's no point having it even visible unless there's a reason to click it.

flukeout commented:
@humphd Yeah, I thought about that as well - seems like a good approach. We'd have to think through some of the cases. For now, let's just get an always-visible button going and we can think through all of the use cases for a 'magic console (tm)'...

**Magic Console**
* Be default, the console is hidden and there isn't an explicit option to toggle
* When the project JS code produces either an error, or a log
  * A "console" button shows up somewhere
  * It indicates how many unseen messages there are
* Clicking the console button shows the console
* If the user dismisses the console, they can only re-open it if there is a new, entry in the console

In addition... we could show the latest entry in the console within the button / UI that opens the full console.

@raygervais Don't worry about these things for now, but curious about your thoughts. Just brainstorming.

raygervais commented:
@flukeout, one idea I'll put into the hat before we start drawing, is the option to permanently show, hide, or dynamically show (being your idea above) the console. This toggle could be accessible from within the settings pane, and by default it only dynamically shows so that users who have no need for the console are not aware of it's existence until they `console.log` something. 

Seeing as this deals with settings and the preference manager @humphd, I'm unsure if my input is valid or logical. Just a thought. I suppose I shouldn't implement the minimized console element till we discuss how this should be handled, correct?

flukeout commented:
@raygervais I think it's safe to implement the console toggle at this point (without any special behaviour) - we'll need some way to open & close it anyway. Is that OK with you? We can layer in some of the special functionality into it later - although there is a chance we could move it up to the editor settings panel in the future.

humphd commented:
Some further thoughts:

* We'd need to make sure if we "don't show until there are logs" that we don't annoy users during a session, and continually hide it with every refresh.  If they opt into seeing it, we should probably leave the button up so they can toggle it.
* What about mobile vs. desktop view?  We should give some thought to how that appears
* Related to the above, should we consider reserving an area at the bottom of the preview that has various tools you can use related to what's happening there?  For example, if the console is going down there, maybe the Inspector tool should too?  Maybe the Mobile vs. Desktop responsive views belong there too?  Maybe all of that stuff is available in a panel that slides up/out from the bottom when you hover over an icon that's pinned in the bottom right?  Just some ideas to trigger @flukeout to make something amazing! :)
* An alternative to what I just said in the previous point, we could also put a button at the top of the Preview where we have all the other tools now--it's pretty similar.  The icon we use could have a badge on it, like an unread notification count, to tell you that you have logs.  The down-side of this is that it's a bit tricky to wire that up in Thimble vs. Brackets, but it's not that hard.

flukeout commented:
As far as the console functionality, I'd also be interested in removing the Errors, Warnings & Logs toggles and just have Errors and Logs on by default. I don't think those distinctions are as important to beginners, and Warnings aren't ever all that useful in my experience.

humphd commented:
Yeah, @raygervais had the same idea when I spoke to him.  I think it would be nice to consider css for each of those, so you can visually pick them out, and we don't have to do exactly what browser dev tools do, either.

flukeout commented:
Lots of great ideas here, but let's scope down what we'd like to see in this first PR

* [ ] Updated console styling including differentiating errors & logs - Luke
* [ ] Add a basic toggle element that shows the console when it's hidden - Ray
* [ ] Enable Errors & Logs, disable Warnings - Ray
* [ ] Change log warnings elements from ``input`` to div, for easier copy&paste - Ray
* [ ] Limit console entries to those produced by user code only - Ray
* [ ] Only show console UI if project has a javascript file - Ray
* [ ] Restore console if it was open when editor is refreshed - Ray

Sound good for Console 1.0 @humphd @raygervais

raygervais commented:
Sounds good to me @flukeout! 

@humphd would I be able to rebase after all the implementation is complete, perhaps with the guidance from a git pro? :smile:

raygervais commented:
@flukeout, still working on the various items. I wanted to show you and @humphd a minor item that I wanted your input on as to why it would be happening. I've updated the log elements from `input` to `div`, except now they are not selectable or focusable. I've already tried messing with Z-Indexes to see if an element was covering the area, but that did not help. Any ideas as I press on?

humphd commented:
Let's rebase this on Friday.  I'll help you when I see you, and we can walk through it together.  Shouldn't be hard.

flukeout commented:
For sure @raygervais - keep me posted when you want me to dive in again, and feel free to check things off that list above as you make the changes!

raygervais commented:
@flukeout, I've completed the following tasks so far, with the exception of the selectable text issue: 
- Console now receives and displays user-relevant data (data fed from the `ConsoleManager.js` file). 
- Errors and Logs have been enabled by default, note that I actually removed the entire filtering process so that we could use the previously implemented `console.*` functions including info, time, timeEnd. This means that instead of buttons, we could simply just display the count for each console type. 

I'll keep working on the other items. Reason why I posted here is because I can't check off items it appears.

raygervais commented:
@flukeout and @humphd, I've added the following features and observations that were requested in the above list. 
- The console now automatically opens when the user's code triggers a `console.*` function if it was closed previously. I'm aware this may be intrusive and wanted to ask when it should specifically always open (in the case where it's closed), example any `console.error` should demand the console be open.  
-  I have in place a small bit of code which can detect that JavaScript exists in a specified project, how would I call the function after the page loads? I tried to call it previously in  the console's `Init`, but the function itself still doesn't see the user's files. Where could I call said function so that it has access to the entire `window.document`? 
```js
    function getScriptExists() {
        let scripts = getElementsByTagName('script');
        if (scripts.length == 0) {
            panel.hide();
        }
    }
```

flukeout commented:
Hey @raygervais - I'll test out the latest changes and get back to you on the icon placement. I like that icon though!

flukeout commented:
Hey @raygervais was just testing this now and noticed that error messages aren't showing up. I have some code like this...

```
console.log("hello world");
conconsole.log("hello world");
```

The second bad ``conconsole.log`` doesn't show an error in our new console.

raygervais commented:
Hey @flukeout, I suppose your issue with the malformed console statements not showing errors in the console itself is because it's being fed data from the `ConsoleManager.js` file, which had only been configured at the time for valid `console.*` functions. I can look into how we could capture the errors which are present in the user's code in the same way. 

> Also, as far as the toggle to open the console goes, can you just append it to the page so it shows up somewhere in the bottom right of the preview? The styling is not important at this point, I'll definitely pitch in there in the near future.

I'll append the toggle at the bottom of the page as requested in upcoming commit. 

> Also, can you make the open console when we see a console.log feature an optional toggle so it's easy to disable or enable in the future in the code?

Regarding the last part, I assume that would work with the `PreferencesManager`? @humphd any input on this one? 

> Thanks so much!

Happy to be contributing, it's been quite the ride so far! 👍

humphd commented:
A few thoughts:

* You can't test for the presence of scripts in the preview file, because the code that you're running to do it is running in a script.  In other words, there is *always* injected script.
* To deal with things like `consoooole.log` and other errors, you probably want to add a global error handler in your remote script, which calls your `console.error`.  Something like this:

```js
// Untested code! see https://developer.mozilla.org/en/docs/Web/API/GlobalEventHandlers/onerror
window.addEventListener("error", function(messageOrEvent, source, lineno, colno, error) {
    // log this info to your console.error somehow.
    // I'm unclear if you get all these args passed to you with addEventListener vs. window.onerror
    // but I'd prefer you did it this way, so we don't make it impossible for user code to add onerror.
}, false);
```

We could also add code later to highlight the code in your editor that caused the error, so you can "live debug" things.  But that can be an enhancement vs. something you have to do here.

* Regarding "only show the console the first time you log something" I think you just need a run-time variable, maybe `var everAutoShown = false;` and the first time you show it, you set it to `true` and subsequent log messages can check the value of this variable and not repeat the process.  It doesn't need to be a pref, I don't think, just something that happens for the lifetime of the app.  We can alter that later.

raygervais commented:
@humphd, after our discussions today regarding the `blob` handling, I implemented this item: 
```js 
    window.addEventListener("error", function(messageOrEvents) {
        var lineNumber = ((messageOrEvents["error"]["lineNumber"] - 2079) > 0)
            ? messageOrEvents["error"]["lineNumber"] - 2079
            : 0;
        var stackTrace = messageOrEvents["error"]["stack"];
        var args =
            ['Message: ' + messageOrEvents["message"] + '\n'
                + 'Line: ' + lineNumber + '\n'
                + 'Column: ' + messageOrEvents["colno"] + '\n'
                + 'File: ',
                stackTrace
            ];

        transportSend("error-handler", args);
    }, false);
```
This, in turn with this code that I wrote  handles the blob: 
```js
if (type === "error-handler") {
            type = "error";

            // Handle Blob URLS
            var stackTrace = args[1].split(",");
            for(var i = 0; i < stackTrace.length; i++) {
                var trace = stackTrace[i].split(":");
                trace = BlobUtils.getFilename("blob:" + trace[1] + ":" + trace[2] + ":" +  trace[3]);
                trace = trace.split("/");
                args[0] = args[0] + trace[trace.length - 1];
            }
            args = args[0];
        }
```
Opinions on how I could optimize the above?

Working on the handling of automatically opening the console now, then @flukeout I'll add the toggle button in Bramble, and the last part I believe will be your styling.

raygervais commented:
Question @flukeout, which element would you like the console toggle to be a appended to? Or, did you have a specific `absolute` style / placement which I could apply if it were appended to the entire window as another element?

humphd commented:
@raygervais what does `stackTrace` look like when you send it?  Can you paste a raw example?

I think you should probably change `PostMessageTransport` to call the code we already have for this, see https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/PostMessageTransport.js#L44-L59

raygervais commented:
@humphd 
`"testerConsole@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:2100:9onclick@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:1:1"`

raygervais commented:
@flukeout, I've added the requested toggle button with some basic styling. Did you want me to add the message count beside it if possible? Also would you like me to make a dark theme?

![image](https://cloud.githubusercontent.com/assets/14265337/24841194/10ac00f0-1d4c-11e7-944e-15dd16c13d03.png)

humphd commented:
@raygervais after you've established that this is a console message request, in your `if` [here](https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/PostMessageTransport.js#L96), make a call to `resolveLinks()`, passing the portion of the `message.data` object contains those blobs, and have it turn them into files.

Also, this business of doing `- 2079` on the line numbers--if you want to do this, you'll first have to write some code in the `PostMessageTransport` that figures out *exactly* how many lines it is programmatically, since we can't assume this number will stay constant.  I'd take the string you are injecting for the remote scripts, and add a function that figures it out based on their contents, minus any HTML lines for `<script>`, etc.

raygervais commented:
@humphd, the blobs as you can see include `functionCall@` before the blob syntax itself, and each blob is preceeded by the line number / column. These items don't allow for `resolveLinks()` to return the corrected URL. 

`ExampleBlob`
```js
 "testerConsole@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:2100:9onclick@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:1:1"
```

Here's the code I'm trying out: 
`PostMessageTransport`
```js
            if(ConsoleManager.isConsoleRequest(msgObj.message)) {
                if(msgObj.data.type === "error-handler"){
                    msgObj.data.args["Stack"] = resolveLinks(msgObj.data.args["Stack"]);
                }
                ConsoleManager.handleConsoleRequest(msgObj.data);
                return;
            }
```

`ConsoleManagerRemote`
```js
    window.addEventListener("error", function(messageOrEvents) {
        var args = {
            "Message": messageOrEvents["message"],
            "Stack": messageOrEvents["error"]["stack"]
        };

        transportSend("error-handler", args);
    }, false);
```

Any ideas as to how to correct this?

humphd commented:
You should probably take the stack you get, split it on `'@'`, and then you can write a regex to break out the 3 parts you need, and run the first (the blob url) through BlobUtils:

```js
// blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:2100:9onclick
var regex = new RegExp('(blob:.+):([^:]+):(.+)', 'gm');
```

Now you can loop through all the separate lines you get when you split on `@` and run each one through the regex, then get the match for the blob url and replace with a filename.

raygervais commented:
@flukeout @humphd The console does not pop up if the user calls a `console.*` function if they had previouslly closed the console using the close button on the right. 

Aside from fixing the error-handling, did you want me to add the `console.*` count to the console toggle similar to your mockup above?

humphd commented:
I think we should drop the count, since we auto-update the live preview a lot more often than the user realizes, and it will cause this to flicker on their screen all the time.  It's probably better to just remain static with the word "Console" or whatever.

raygervais commented:
@flukeout, you'll see that I added some styling to the console, and I think I have an idea as to how we can make the console both dark & light themed .. putting the styling itself in the `~/src/extensions/default/bramble/stylesheets/*Theme.css` files which will take attributes such as background, font colors, etc. 

As you can see from the screenshot, I've removed the buttons since we decided not to filter anymore, and instead replaced them with a standard readout & badge. I wanted to ask you if you had any preference as to the styling of the badges themselves. 

Regarding color, I was going to use the same editor colors where appropriate for the various console output, does that sound wise? 

![image](https://cloud.githubusercontent.com/assets/14265337/24971352/b1f105ec-1f85-11e7-8a2c-a6262a92530c.png)

@humphd Noted, will finish the error-handling then.

raygervais commented:
@flukeout & @humphd, I've completed the vast majority of the error handling items, and wanted to get your opinion before I sign off on that section of the console. 

When displaying the stack trace, how would you like the content displayed? 
The current displaying of error messages looks like the following, with the addition of that `,` that I don't know how to remove or how it's getting there. Will look into further while I await your responses. 

![image](https://cloud.githubusercontent.com/assets/14265337/24986127/63a082d2-1fc6-11e7-9aaa-38c4608ea7d8.png)

Firefox displays console errors this way: 
![image](https://cloud.githubusercontent.com/assets/14265337/24986209/d0b2a5bc-1fc6-11e7-8fc1-4f9cc73f5abb.png)

Jetbrain's IntelliJ 
**Far too complex I'm aware, simply asking for how you'd like data to be presented while providing examples**
![image](https://cloud.githubusercontent.com/assets/14265337/24986363/cd47fe80-1fc7-11e7-9d30-5da2d0d4cfe9.png)

Firebug
![image](https://cloud.githubusercontent.com/assets/14265337/24986512/b082e00c-1fc8-11e7-9a5f-aed8cf691787.png)

As you can see, the messageOrEvent variable provides quite a bit to work with if you wanted more data. I hadn't explored too deeply into the object, so I cannot say for certain that it contains anymore useful content.

![image](https://cloud.githubusercontent.com/assets/14265337/24987235/d1f3c1d0-1fcc-11e7-9f54-ddce8585e786.png)

humphd commented on src/extensions/default/bramble/htmlContent/console.html:
```diff
@@ -0,0 +1,20 @@
+        <div id="filterBtns" class="btn-group margin-left-small">
+            <div id="btnErrors" class="btn smallgit " data-filter="error">
+                Errors <span class="badge">0</span>
+			</div>
```
Some weird indenting here

humphd commented on src/extensions/default/bramble/htmlContent/console.html:
```diff
@@ -0,0 +1,20 @@
+            </div>
+            <div id="btnLogs" class="btn small" data-filter="log">
+                Logs <span class="badge">0</span>
+			</div>
```
And here.

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,152 @@
+
+    var panel
+        , icon = null
+        , panelHTML = require("text!htmlContent/console.html");
```
I know you didn't write this, but can you switch these to trailing commas vs. prefixing?

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,152 @@
+        , icon = null
+        , panelHTML = require("text!htmlContent/console.html");
+
+    icon = $("<button class=\"editor-console-icon-indicator\" title=\"Click to toggle the JavaScript Console \">Console</button>");
```
This is no good because we can't localize it.  We need to get the string from the `Strings` object.

raygervais replied:
Should I remove? If not, how does one access the strings object?

humphd replied:
You will need to render this with `Mustache` and inject the string, something like https://github.com/mozilla/brackets/commit/5158dfb948088767219ba88d39b3295b14bb9264.  NOTE: we use the `brackets.getModule("Strings")` module vs. `require('strings')` for the extension.  Most extension authors are encourage to provide their own strings.  We do it differently, and put our strings into the main bracket's `Strings` module.  You'll eventually have to have the string go in Thimble's files, which is where it gets translated.  @gideonthomas can help you when you get there.

raygervais replied:
@gideonthomas @humphd, I've added two strings to the `locales/en-US` file: 
```
CONSOLE_TOGGLE_TITLE=Console
CONSOLE_CLEAR=Clear
```
Would I make these `IDs` in the HTML content? How do I inject them into the code? Thanks in advanced!

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,152 @@
+            , $element = ""
+            , countBtns = panel.$panel.find(".badge")
+            , data = logData
+            , i = 0;
```
Same thing here, trailing commas, please.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -15,8 +17,17 @@ define(function (require, exports, module) {
-        // TODO: Show this in Custom Console UI, see issue #1675 in Thimble
+        if (type === "error-handler") {
+            type = "error";
+			args[0] = args["Stack"];
```
Indenting.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -21,7 +21,18 @@
+        var args = {
+            "Message": messageOrEvents["message"],
+            "Line": messageOrEvents["error"]["lineNumber"],
+            "Stack": messageOrEvents["error"]["stack"]
```
Can you make all these keys lowercase?  stack vs. Stack, etc.

humphd commented on src/extensions/default/bramble/lib/PostMessageTransport.js:
```diff
@@ -93,11 +93,25 @@ define(function (require, exports, module) {
+                        var stackItem = stackTrace[i].match(regex);
+                        stackItem = (BlobUtils.getFilename(stackItem[0].replace(endingRegex, ''))).split('/');
+                        msgObj.data.args.push('Function Line:' + stackItem[1] + ' File:' + stackItem[stackItem.length -1 ]);
+                    }
```
Move all this code for regex'ing the message into your `ConsoleManager` module.  Just call `ConsoleManager.handleConsoleRequest(msgObj.data);` and then do it in there (or call a function in there).  Let's keep this file about routing messages from the iframe.

humphd commented on src/extensions/default/bramble/lib/UI.js:
```diff
@@ -133,7 +133,7 @@ define(function (require, exports, module) {
             ViewCommandHandlers.setFontSize(fontSize);
         }
-
+		
```
Revert this file change: `git checkout master src/extensions/default/bramble/lib/UI.js `

humphd replied:
You missed this, please do it before next review.

humphd replied:
This is still needing to be done.  Why do you keep skipping it?

humphd commented on src/extensions/default/bramble/main.js:
```diff
@@ -32,7 +32,8 @@ define(function (require, exports, module) {
         RemoteCommandHandler = require("lib/RemoteCommandHandler"),
-        RemoteEvents         = require("lib/RemoteEvents");
+        RemoteEvents         = require("lib/RemoteEvents"),
+		ConsoleInterface     = require("lib/ConsoleInterfaceManager");
```
Indenting

humphd commented on src/extensions/default/bramble/stylesheets/consoleTheme.css:
```diff
@@ -0,0 +1,75 @@
+    width: 100%;
+    margin: 5px 10px;
+    background: none;
+    border: 0px;
```
s/0px/0/ (no units on 0)

humphd commented on src/extensions/default/bramble/stylesheets/consoleTheme.css:
```diff
@@ -0,0 +1,75 @@
+
+#editor-console #filterBtns .btn {
+    border-radius: 0;
+	background: #222;
```
Indenting

humphd commented on src/extensions/default/bramble/stylesheets/consoleTheme.css:
```diff
@@ -0,0 +1,75 @@
+    position: absolute;
+    width:120px;
+    background: #eee;
+	bottom: 2px;
```
Indenting.

raygervais commented:
@humphd I've fixed all the items you wanted changed except the localization, which I will get to later this week. When you get a chance, could you give it another review, or would you rather wait for the localization to be implemented? 

PS. indenting was artifact of brand-new atom install which wasn't configured for spaces... Will keep that in mind for future contributions 👍

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,150 @@
+        $(countBtns[2]).html(count(logData, 'log'));
+    }
+
+	function render() {
```
Indenting

humphd commented on src/extensions/default/bramble/htmlContent/console.html:
```diff
@@ -0,0 +1,20 @@
+<div id="editor-console" class="bottom-panel vert-resizable top-resizer">
+    <div class="toolbar simple-toolbar-layout mainToolbar">
+        <span class="title">Console</span>
```
Let's also remember that the strings in this HTML are also going to need to get inserted via localization.

humphd replied:
What's your plan here with localization, still hasn't been addressed.

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,150 @@
+
+    var wasClosedByUser = false;
+
+
```
Remove extra blank line.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -15,8 +17,30 @@ define(function (require, exports, module) {
+            // Handle Blob URLs
+            for(var i = 1; i < stackTrace.length; i++){
+                var stackItem = stackTrace[i].match(regex);
+                stackItem = (BlobUtils.getFilename(stackItem[0].replace(endingRegex, ''))).split('/');
```
Why do you split here?  Are you trying to get the basename on its own, separate from the path?  You should use the `Path` module on `Filer` if so, and do `Path.basename(...)`.

raygervais replied:
` "testerConsole@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:2100:9onclick@blob:http://localhost:8000/4eaede7e-49a6-4c1e-997f-3da508f77cfa:1:1"`

Splitting to isolate the blob URL from the function call. 

By calling `Path.basename(...)`, what does the function variable return?

humphd replied:
I don't follow.  The only `/`s I see in that string are within the blob URLs themselves.  Can you clarify what you're a) passing in as your string; b) what you want to get out?  Sorry for being confusing, I'm just trying to understand your goal vs. your method.

raygervais replied:
Ah, my mistake in the previous response. I thought you were asking why I was splitting on the `@` symbol (browser window did not show the split at the end of my line you're referring too. 

By not splitting, I'm `stackItem` is assigned the following: 
```js
/7/projects/30/index.html
```

I believe the reason I wanted to split was to separate the information itself. If you prefer I'll remove the `split` function and adjust the output accordingly to display the above. Apologies again for the confusion.

humphd replied:
OK, that's what I thought.  For sure you want to use `Path.basename()` here, it will give you `index.html` for that path.

raygervais replied:
Forgive for the questions relating to `Path`, just want to ask before implementing to avoid delays. 
I'd use `Path.basename` instead of `BlobUtils.getFileName`? Or am I looking at this wrong?

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -15,8 +17,30 @@ define(function (require, exports, module) {
+            args[0] = type + ": " + args[0];
+        }
+
         console[type].apply(console, args);
```
Why do you do both of these next two lines?  Isn't adding enough?

raygervais replied:
Just for debugging, I can remove the `console[type].apply(console, args)` line since it's no longer needed.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -21,7 +21,18 @@
+        var args = {
+            "message": messageOrEvents["message"],
+            "line": messageOrEvents["error"]["lineNumber"],
+            "stack": messageOrEvents["error"]["stack"]
```
NOTE: in an object literal (vs. JSON), when dealing with keys as `String`, you don't need to quote.  Also, I think we should be careful not to crash if some browser doesn't provide the correct things:

```js
var message = messageOrEvents.message;
var error = messageOrEvents.error || {};
var args = {
    message: message,
    line: error.lineNumber,
    stack: error.stack
};
```

humphd commented:
No, you'd still use `BlobUtils.getFileName`, but if you want to then show only the last part of the filename, you can get it with `Path.basename`.

flukeout commented:
@raygervais is there an email I can reach you at? I'm collecting information for a contributors page on the Thimble website and I'd like to include you. Can you please email me at ``luke@mozillafoundation.org``? Thanks!

raygervais commented:
@humphd, I've incorperated `Path.basename` into the blob handing logic. Is there anyway to remove the `,` that is found at the start of each line? 

![image](https://cloud.githubusercontent.com/assets/14265337/25109390/2926a770-23aa-11e7-805b-b7f1f6b09bed.png)

@flukeout 
I've made the console inherit the background color elements for the dark and light themes, can you confirm it works on light the light theme? Also do you have a color which works well for both themes for the `console.log` function? I originally had it using `#FFFFFF`, but alas that is only suitable for the dark theme. 

![image](https://cloud.githubusercontent.com/assets/14265337/25109446/8b5160b6-23aa-11e7-94d6-e3c464699857.png)

humphd commented:
I tried this out locally.  Some things I notice:

![screen shot 2017-04-18 at 10 53 18 am](https://cloud.githubusercontent.com/assets/427398/25137283/531a5880-2425-11e7-96fc-d3c89b4338ae.png)

* The console isn't scrollable.  What you're doing with `overflow` doesn't seem like it's working.  I think you need to use `overflow-y: scroll` on the `.console` class.
* The text in the console isn't selectable.  Change `user-select` to `text` for the `.console` class, and it will work
* Errors are counted, but not always shown; sometimes there are errors added with no content.  Probably you shouldn't add new items to the console if the content is `null`, `undefined`, `""`, etc.
* You should probably use https://developer.mozilla.org/en/docs/Web/API/Element/scrollIntoView to force the newly added `<div>` in the console to always show.  It's odd that it can't be seen when I have the console open.  I'd expect to be able to always see it.
* I feel like the counts should be removed. these all look like buttons, but they do nothing.  Why do we have them?  I'm not sure they add anything:

![screen shot 2017-04-18 at 11 18 46 am](https://cloud.githubusercontent.com/assets/427398/25138582/e7efe0bc-2428-11e7-84f8-00c11a17b8e7.png)

* `Error` objects aren't being shown properly:

![screen shot 2017-04-18 at 11 21 44 am](https://cloud.githubusercontent.com/assets/427398/25138726/4a918f40-2429-11e7-9740-a51a45088e7a.png)

* You're splitting on `@` but the stacks don't include them (at least in Chrome):

![screen shot 2017-04-18 at 11 26 09 am](https://cloud.githubusercontent.com/assets/427398/25138947/e6cbfb8e-2429-11e7-88fc-aff7298c3fad.png)

Your code for unwinding the stack and printing it seems wrong to me.  Consider a stack string that looks like this:

```
"Error: error
    at log (blob:http://localhost:8000/a9ca17d8-d171-43b4-bb35-da5d0a1fbd0c:2098:15)
    at HTMLButtonElement.onclick (blob:http://localhost:8000/a9ca17d8-d171-43b4-bb35-da5d0a1fbd0c:2089:160)"
```

All you need to do with this is to find every URL and rewrite it to a filename.  So you need to extract all the `blob:...` URLs and run them through `BlobUtils.getFilename()`, which you do.  That will give you back an absolute path, and you might choose to then strip it down to just the basename.

If this is going to block us landing this bug, I'd suggest we hive off this portion and do it in a new bug after this lands.

In general, let's start figuring out what should happen in this bug vs. follow-ups.  @flukeout, I also think the UI needs a bit of work, have you reviewed it yet?

flukeout commented:
Hey @raygervais once you address @humphd's feedback, I'll try it out and provide styling updates! Please cc me when you've done so - thanks!

humphd commented on src/extensions/default/bramble/stylesheets/consoleTheme.css:
```diff
@@ -23,10 +23,10 @@
-
-#editor-console .console .log {
-    color: #FFFFFF;
+    user-select: text;
```
With CSS prefixes, you always want to put the unprefixed version last in the order.  You do this because a browser might support both `-moz-user-select` and `user-select`, and eventually only the latter.  You want it to use the unprefixed if possible.

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -69,10 +68,6 @@ define(function (require, exports, module) {
         var countBtns = panel.$panel.find(".badge");
```
Does this name still make sense, `countBtns` if we don't have them?

raygervais commented:
@humphd, I've been experimenting with how to handle object arguments such as your example of `console.error(new Error("Test"));`. 

Here's the code I have so far, but for the life of me the code does not access the object's variables itself. Any ideas? If you do think this is going to take too long, I can expand upon it after this is landed. An optimization I'd do to the code below is remove the `arg` variable entirely, which was used for a `forEach` loop styled iterator in a previous idea. 
```js
["log",
     "warn",
     "info",
     "debug",
     "info",
     "error",
     "clear",
     "time",
     "timeEnd"].forEach(function(type) {
        console[type] = function() {            
            var values = [];
            var args = [];
            var arg;

            // Handle Object Arguments
            for(var i = 0; i < arguments.length; i++) {
                if(typeof(arguments[i]) === 'object' || typeof(arguments[i] === 'function')) {
                    for (var key in arguments[i]) {
                        values.push(arguments[i][key]);
                    }
                    arg = values;
                    values = [];
                } else {
                    arg = arguments[i];
                }
                
                args.push(arg);
            }
            transportSend(type, args);
        };
    });
```

Looking into your notes on localization now, along with how to properly display the stack trace mentioned earlier in this thread.

humphd commented:
In JS `arguments` is kind of a special thing.  It's not really an array, though you can iterate through it, and check its length.  You basically need to turn it into an Array using `var args = Array.prototype.slice.call(arguments);`  Then you don't need to iterate through it.

raygervais commented:
That is what I had done previously, but I noticed that if you had an Object in it such as `Error`, the object itself would become `{}` with no values inside it. If that makes sense.

humphd commented:
You can special-case a check for `Error` objects by doing `if(args[0] instanceof Error){...}` and then pull off whatever you need.  I don't think you need to care too much about checking for all kinds of edge cases, but probably worth it with `Error`.

raygervais commented:
Will do, thank you. Also testing Localization now.

gideonthomas commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,137 @@
+
+        // Localization
+        panelHTML = Mustache.render(panelHTML, {"Strings": Strings});
+        icon = Mustache.render(icon, {"Strings": Strings});
```
You cannot pass anything other than a string to `Mustache.render` as the first param. In this case, `icon` is a jquery object.

humphd commented on locales/en-US/editor.properties:
```diff
@@ -193,3 +193,8 @@ ERROR_MOVING_FILE_DIALOG_HEADER=Move Error
+# extensions/default/bramble/console
+CONSOLE_TITLE=CONSOLE
+CONSOLE_TOOLTIP=Click to use the JavaScript Console.
+CONSOLE_CLEAR=Clear 
```
New line at end of file

raygervais replied:
Fixed!

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,131 @@
+        panelHTML = require("text!htmlContent/console.html");
+    
+    icon = $("<button class=\"editor-console-icon-indicator\" title='{{CONSOLE_TOOLTIP}}'>{{CONSOLE_TITLE}}</button>");
+    icon.appendTo($("#editor-holder"));
```
Move this into `htmlReady` below.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -15,8 +18,31 @@ define(function (require, exports, module) {
+            args[0] = type + ": " + args[0];
+        }
+
+        if (args !== undefined){
```
`if (args) {`

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -16,12 +16,40 @@
+        var args = {
+            "message": message,
+            "line": line,
+            "stack": stack
```
Drop the `"..."` quotes from these keys, you only need that in JSON.

humphd commented on src/extensions/default/bramble/main.js:
```diff
@@ -32,7 +32,8 @@ define(function (require, exports, module) {
         RemoteCommandHandler = require("lib/RemoteCommandHandler"),
-        RemoteEvents         = require("lib/RemoteEvents");
+        RemoteEvents         = require("lib/RemoteEvents"),
+        ConsoleInterface     = require("lib/ConsoleInterfaceManager");
```
Why in `main.js` vs. where it's used?

raygervais commented:
@humphd apologies for the delay, I've fixed the stylistic issues you've brought up and have been working on localization, which seems to result in a `too many recursion` error at run time. I see that because of the change in OS, some of the git history has become cluttered so I'll be sure to `rebase` in the end. 

Would you or @gideonthomas be able to look at my recent commit and help determine the recursion issue?

humphd commented:
You have a load timing error here.  You're loading `ConsoleInterfaceManager.js` too early in `src/extensions/default/bramble/main.js`, and the `Strings` module hasn't loaded yet.  Don't require that file here, since you don't use it.

gideonthomas commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -16,12 +16,40 @@
-            var args = Array.prototype.slice.call(arguments);
+        console[type] = function() {            
+            var values = [];
+            var args = Array.prototype.slice.call(arguments); 
```
Let's simplify this logic to:
```js
var args = Array.from(arguments).map(function(arg) {
  if(!(arg instanceof Error)) {
    return [];
  }

  return Object.keys(arg).map(function(key) {
    return arg[key];
  });
});
```

gideonthomas replied:
Ah yeah, IE 11. Instead of the `Array.from(arguments).map` just use `Array.prototype.map.call(arguments, function(arg) { ... });`

gideonthomas commented:
@raygervais I think you might need to rebase your patch on mozilla/master

humphd commented:
Be careful of IE11 js incompatibility here.

On Apr 25, 2017 4:19 PM, Gideon Thomas <notifications@github.com> wrote:

@gideonthomas commented on this pull request.

________________________________

In src/extensions/default/bramble/lib/ConsoleManagerRemote.js<https://github.com/mozilla/brackets/pull/672#discussion_r113299629>:

> @@ -16,12 +16,40 @@
      "clear",
      "time",
      "timeEnd"].forEach(function(type) {
-        console[type] = function() {
-            var args = Array.prototype.slice.call(arguments);
+        console[type] = function() {
+            var values = [];
+            var args = Array.prototype.slice.call(arguments);


Let's simplify this logic to:

var args = Array.from(arguments).map(function(arg) {
  if(!(arg instanceof Error)) {
    return [];
  }

  return Object.keys(arg).map(function(key) {
    return arg[key];
  });
});

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mozilla/brackets/pull/672#pullrequestreview-34675954>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAaFhm3L9JxX0o3VJfQP60SR7tLtVu3tks5rzlVRgaJpZM4MtYLn>.

raygervais commented:
@humphd, where would you suggest I load the file? I was looking at adding it at the end of `src/extension/default/bramble/main.js`

```js
function init(e) {
    ...   
}
```

Having tested this, it appears that the `recursion` issue is still present.

humphd commented on src/extensions/default/bramble/lib/ConsoleInterfaceManager.js:
```diff
@@ -0,0 +1,130 @@
+        ExtensionUtils = brackets.getModule("utils/ExtensionUtils"),
+        PreferencesManager = brackets.getModule("preferences/PreferencesManager"),
+        Mustache = brackets.getModule("thirdparty/mustache/mustache"),
+        Strings = brackets.getModule("Strings");
```
I think this should be `brackets.getModule("strings");` with a lower-case `s`.  Try that.

raygervais replied:
That fixed the recursion issue!! I'll commit the current version and will keep working on `Mustache`, which still doesn't see my `Strings` that I added to `locale/en-US/editor.properties`.

flukeout commented:
Hey @gideonthomas and @raygervais - just checking in on how this is going overall. I'm going to test it out now, let me know what the status is.

raygervais commented:
@flukeout, I've been meaning to get back to this (I last left off with the localization not working), but I've been in the middle of a move. I will try to resume more work on this later on this week. Apologies for going dark for almost a week and then some.

flukeout commented:
Thanks for the update @raygervais - this is one of the features we really want to get launched on May 17th, so we'll definitely be helping out & assisting. Hope the move went/is going well!

raygervais commented:
Furthermore I believe the `Error` object & `window.error` stacktrace isn't complete, this could be implemented after release though.

flukeout commented:
Okay, got it up and running on my local machine. I'll work on some style changes and submit a PR against your working branch sometime today or tomorrow @raygervais

flukeout commented:
Yeah, I think if we have just a basic feature set that allows console.logs to come through that would be great. Particularly since we're launching the Snippets feature too and that includes some ``console.log`` calls.

raygervais commented:
Noted, I'd be happy to keep contributing after the initial release and flesh out the console. I can try later tonight to remove any non-functional code such as the error handler for the time being, or comment out if that's allowed in this instance.

flukeout commented:
@raygervais One of the things I've found while testing / styling is that it doesn't escape html tags. If I  ask it to ``console.log("</div>")`` then it actually puts in a closing div tag into the console markup. Like this...

![image](https://cloud.githubusercontent.com/assets/25212/25727254/1b40a026-30dd-11e7-832e-5318bc5a2159.png)

flukeout commented:
Hey @raygervais another thing I noticed is that it removes all of the divs in the console and adds them all again every time there's a new log. Is it possible to just append the new <div> elements as they are generated? This would allow us to add a little effect to newly added logs to make them more noticeable.

flukeout commented:
Okay, so I reworked it a bit so that we don't keep an object of the logs, we just append the latest ones when we get them. This way we can add a little animation effect to the new ones that makes the easier to notice...

![_console_strobe](https://cloud.githubusercontent.com/assets/25212/25727931/c4dcd160-30e0-11e7-9369-195f9d666ac3.gif)

I'll submit these changes in a PR so you can review them @raygervais

flukeout commented:
Hey @raygervais - I submitted a PR against your branch.

flukeout commented:
Weird, it's like I never made the PR - looking into it.

flukeout commented:
Okay phew, the PR against this branch is here @raygervais https://github.com/raygervais/brackets/pull/3

raygervais commented:
@flukeout thanks so much for the help & styling! Looking at the code right now.

humphd commented:
Looks like `src/extensions/default/bramble/htmlContent/console-icon-on.png` and `src/extensions/default/bramble/htmlContent/console-icon.png` are unused in this PR, let's remove those files.

Can someone update me on what needs to happen on this before it can land?  Does it need another review?  I'd love to get it finished within the week, if possible, and am happy to help make that happen.

gideonthomas commented:
This just needs a final review + test (I think)

flukeout commented:
Re-testing now since @raygervais merged my PR into his branch.

flukeout commented:
Yeah, seems like it works - let's give it a review and merge if we can.

humphd commented:
OK, I'll give it a spin and review this aft or tomorrow morning.

raygervais commented:
@humphd @flukeout I've fixed a merge conflict relating to `UI.js`, I'll give it a spin as well tonight.

humphd commented:
I spent the morning on this, and in order to expedite landing this, I decided to just write my changes vs. making comments.

Building on the latest commit (ab6f2a6) I've done the following work:

https://github.com/raygervais/brackets/compare/issue-1675...humphd:issue-1675

@gideonthomas, @raygervais can you take a quick look and let me know if you see any problems?  If not, I'm going to push all this.

gideonthomas commented:
Nice work! I give this an r+

raygervais commented:
Tested, works! Thanks @humphd for polishing & resolving some of the issues.

humphd commented:
Excellent.  I'm going to land this in https://github.com/mozilla/brackets/pull/752 and close this.  Amazing work, @raygervais!  Thank you for sticking with it for so long.

flukeout commented:
Yeah, thanks @raygervais - awesome feature! 👍

raygervais commented:
@humphd just doing some more testing here since you've had to revert a couple items. It no longer works on any workstation, and gets caught in `dist/config.js` every time. What is the correct way to test, because I must be doing something wrong. 😞
